File: jarboot-api/src/main/java/com/mz/jarboot/api/service/ServiceManager.java
package com.mz.jarboot.api.service;

import com.mz.jarboot.api.constant.TaskLifecycle;
import com.mz.jarboot.api.event.Subscriber;
import com.mz.jarboot.api.event.TaskLifecycleEvent;
import com.mz.jarboot.api.pojo.JvmProcess;
import com.mz.jarboot.api.pojo.ServiceInstance;
import com.mz.jarboot.api.pojo.ServiceSetting;

import java.util.List;

/**
 * 服务管理
 * @author majianzheng
 */
public interface ServiceManager {

    /**
     * 获取服务列表
     * @return 服务列表
     */
    List<ServiceInstance> getServiceList();

    /**
     * 获取服务信息
     * @param serviceName 服务名称
     * @return 服务信息 {@link ServiceInstance}
     */
    ServiceInstance getService(String serviceName);

    /**
     * 一键重启，杀死所有服务进程，根据依赖重启
     */
    void oneClickRestart();

    /**
     * 一键启动，根据依赖重启
     */
    void oneClickStart();

    /**
     * 一键停止，杀死所有服务进程
     */
    void oneClickStop();

    /**
     * 启动服务
     * @param serviceNames 服务列表
     */
    void startService(List<String> serviceNames);

    /**
     * 停止服务
     * @param serviceNames 服务列表，字符串格式：服务path
     */
    void stopService(List<String> serviceNames);

    /**
     * 重启服务
     * @param serviceNames 服务列表，字符串格式：服务path
     */
    void restartService(List<String> serviceNames);

    /**
     * 通过服务配置启动服务
     * @param setting 服务配置
     */
    void startSingleService(ServiceSetting setting);

    /**
     * 获取未被服务管理的JVM进程信息
     * @return jvm进程信息
     */
    List<JvmProcess> getJvmProcesses();

    /**
     * attach到指定的进程
     * @param pid 进程pid
     */
    void attach(String pid);

    /**
     * 删除服务
     * @param serviceName 服务名
     */
    void deleteService(String serviceName);

    /**
     * 注册事件处理
     * @param serviceName 服务名称
     * @param lifecycle 任务生命周期 {@link TaskLifecycle}
     * @param subscriber 任务处理 {@link Subscriber}
     */
    void registerSubscriber(String serviceName,
                            TaskLifecycle lifecycle,
                            Subscriber<TaskLifecycleEvent> subscriber);

    /**
     * 反注册事件处理
     * @param serviceName 服务名称
     * @param lifecycle 任务生命周期 {@link TaskLifecycle}
     * @param subscriber 任务处理 {@link Subscriber}
     */
    void deregisterSubscriber(String serviceName,
                              TaskLifecycle lifecycle,
                              Subscriber<TaskLifecycleEvent> subscriber);
}

File: jarboot-ui/src/services/ServiceManager.ts
import Request from "@/common/Request";
import CommonNotice from "@/common/CommonNotice";
import {requestFinishCallback} from "@/common/CommonConst";
import StringUtil from "@/common/StringUtil";
import React from "react";

const urlBase = "/api/jarboot/services";

interface TreeNode {
    sid: string;
    name: string;
    title: string | React.ReactNode;
    key: string;
    isLeaf?: boolean;
    children?: TreeNode[];
    icon?: React.ReactNode;
    selectable?: boolean;
}

interface ServiceInstance extends TreeNode {
    status: string;
    group: string;
    path: string;
}

interface JvmProcess extends TreeNode {
    fullName?: string;
    pid: number;
    attached: boolean;
    remote: string;
    attaching: boolean;
    trusted: boolean;
}

export { ServiceInstance, JvmProcess, TreeNode };
/**
 * 服务管理
 */
export default class ServiceManager {

    /**
     * 获取服务列表
     * @param callback
     */
    public static getServiceList(callback: any) {
        Request.get(urlBase, {})
            .then(callback)
            .catch(CommonNotice.errorFormatted);
    }

    /**
     * 启动服务
     * @param services
     * @param callback
     */
    public static startService(services: ServiceInstance[], callback: any) {
        const param = ServiceManager.parseParam(services);
        Request.post(`${urlBase}/startService`, param).then(callback).catch(CommonNotice.errorFormatted);
    }

    /**
     * 终止服务
     * @param services
     * @param callback
     */
    public static stopService(services: ServiceInstance[], callback: any) {
        const param = ServiceManager.parseParam(services);
        Request.post(`${urlBase}/stopService`, param).then(callback).catch(CommonNotice.errorFormatted);
    }

    /**
     * 重启服务
     * @param services
     * @param callback
     */
    public static restartService(services: ServiceInstance[], callback: any) {
        const param = ServiceManager.parseParam(services);
        Request.post(`${urlBase}/restartService`, param).then(callback).catch(CommonNotice.errorFormatted);
    }

    /**
     * 一键重启
     */
    public static oneClickRestart() {
        Request.get(`${urlBase}/oneClickRestart`, {}
        ).then(requestFinishCallback).catch(CommonNotice.errorFormatted);
    }

    /**
     * 一键启动
     */
    public static oneClickStart() {
        Request.get(`${urlBase}/oneClickStart`, {}
        ).then(requestFinishCallback).catch(CommonNotice.errorFormatted);
    }

    /**
     * 一键停止
     */
    public static oneClickStop() {
        Request.get(`${urlBase}/oneClickStop`, {}
        ).then(requestFinishCallback).catch(CommonNotice.errorFormatted);
    }

    /**
     * base64编码
     * @param data
     */
    public static base64Encoder(data: string) {
        if (StringUtil.isEmpty(data)) {
            return Promise.resolve();
        }
        return Request.get(`${urlBase}/base64Encoder`, {data});
    }

    /**
     * 获取未被服务管理的JVM进程信息
     * @param callback
     */
    public static getJvmProcesses(callback: any) {
        Request.get(`${urlBase}/jvmProcesses`, {})
            .then(callback)
            .catch(CommonNotice.errorFormatted);
    }

    /**
     * attach进程
     * @param pid pid
     */
    public static attach(pid: number) {
        return Request.get(`${urlBase}/attach`, {pid});
    }

    /**
     * 删除服务
     * @param serviceName 服务名
     */
    public static deleteService(serviceName: string) {
        const form = new FormData();
        form.append('serviceName', serviceName);
        return Request.delete(`${urlBase}/service`, form);
    }

    private static parseParam(services: ServiceInstance[]): string[] {
        const set = new Set<string>();
        services.forEach(value => {
            if (value.isLeaf) {
                set.add(value.name as string);
                return;
            }
            const children = value.children as ServiceInstance[];
            children?.length && children.forEach(child => set.add(child.name as string));
        });
        return Array.from(set);
    }
}
